backend/src/lib.rs

```diff
@@ -1,104 +1,176 @@
 
-#[cfg(feature = "halo2")]
-impl ProverWithParams for Halo2Backend {
-    fn prove<T: FieldElement, R: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof> {
-        Some(halo2::prove_ast_read_params(pil, fixed, witness, params))
+/// Factory implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F> + 'static> BackendFactory<F> for WithSetupFactory<B> {
+    fn new(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
+        Box::new(ConcreteBackendWithSetup(B::new(degree)))
     }
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params {
-        halo2::generate_params::<T>(size)
+    fn new_from_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
+        Ok(Box::new(ConcreteBackendWithSetup(B::new_from_setup(
+            input,
+        )?)))
     }
 }
 
-#[cfg(feature = "halo2")]
-impl ProverWithoutParams for Halo2MockBackend {
-    fn prove<T: FieldElement>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-    ) -> Option<Proof> {
-        halo2::mock_prove(pil, fixed, witness);
-        None
+/// Concrete dynamic dispatch Backend object, for backends with setup.
+struct ConcreteBackendWithSetup<B>(B);
+
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F>> Backend<F> for ConcreteBackendWithSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        self.0.prove(pil, fixed, witness, prev_proof, output_dir)
     }
-}
 
-#[cfg(feature = "halo2")]
-impl ProverAggregationWithParams for Halo2AggregationBackend {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof {
-        halo2::prove_aggr_read_proof_params(pil, fixed, witness, proof, params)
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error> {
+        Ok(self.0.write_setup(output)?)
     }
 }
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("input/output error")]
+    IO(#[from] std::io::Error),
+    #[error("the backend has not setup operations")]
+    NoSetupAvailable,
+}
+
+pub type Proof = Vec<u8>;
+
+/// Dynamic interface for a backend.
+pub trait Backend<F: FieldElement> {
+    /// Perform the backend proving. If prev_proof is provided, proof
+    /// aggregation is performed.
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()>;
```

Comment:
should return proof?

Reply:
See this: https://github.com/powdr-labs/powdr/pull/417#discussion_r1282218718

Reply:
Actually thinking about it now I strongly disagree with it. The client should receive the proof and decide what to do with it. The way we're doing this now, no one will be able to use the backend crate via powdr for any needs if the proofs always get written to disk, especially in cases of recursion.

Reply:
I think it should return option proof

Reply:
done